# Examples throughout the lesson

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a dataframe
df_test=pd.DataFrame()

In [3]:
# Create a mixed type column
df_test['mix']=['a','b', 1, True]

In [5]:
# Review the result
df_test

mix
0     a
1     b
2     1
3  True

In [7]:
# Check if the column contains different data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
# Change the mixed-data-type column to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Import productsl.csv
path = r'/Users/elvachen/Documents/DA Learning/Immersion-data resources'
df_prods = pd.read_csv(os.path.join(path, 'original data', 'products.csv'),index_col = False)

In [11]:
# Look for the null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create a dataframe to contain the null values
df_nan=df_prods[df_prods['product_name'].isnull()== True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Check the shape of the dataframe
df_prods.shape

(49693, 5)

In [14]:
# Create a clean dataframe without the null-value rows
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# Check the shape of the clean dataframe
df_prods_clean.shape

(49677, 5)

In [16]:
# Look for duplicates 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
# Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
# Check the shape of the new dataframe
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
# Export the clean dataframe to csv file
df_prods_clean_no_dups.to_csv(os.path.join(path, 'prepared data', 'products_clean.csv'))

# Task of 4.5

In [21]:
# 4.5.2 Anything about the data looks off or should be investigated further.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum price is unusual which is 99999 when the 3/4 quartile is 11.2. 

In [28]:
# 4.2.3 Check for mixed-type data in your df_ords dataframe.
df_order_wrangled = pd.read_csv(os.path.join(path, 'prepared data', 'orders_wrangled.csv'),index_col = False)
for col in df_order_wrangled.columns.tolist():
    weird = (df_order_wrangled[[col]].applymap(type) != df_order_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_order_wrangled[weird]) > 0:
        print (col)

In [ ]:
# 4.2.4 If you find mixed-type data, fix it.

There is no mixed-type data found.

In [30]:
# 4.2.5 Run a check for missing values in your df_ords dataframe.
df_order_wrangled.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_sequency                 0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

According to the result above, there is only one column "dats_since_prior_order" that has missing data. The number of missing values in the column is 206209. It makes sences customers who just placed only one order haven't no values for "days_since_prior_order". 

In [ ]:
# 4.2.6 Address the missing values using an appropriate method.

To check if all customers who placed more than one order have respectively values in the column "days_since_prior_order", if not, fixed them with calculation of the durations between appropriate oreder dates. Then, leave the missing values in the column for all first order of the customers in the dataframe, as they interprets correct meaning of the business.

In [34]:
# 4.2.7 Run a check for duplicate values in your df_ords data.
df_order_wrangled[df_order_wrangled.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_sequency, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

According to the above result, no duplicated rows found.

In [ ]:
# 4.2.8 Address the duplicates using an appropriate method.

No duplicates were found in the dataframe.

In [35]:
# 4.2.9 Export your final, cleaned df_prods and df_ords data as ".csf" files in your "Prepared Data" folder.
df_order_wrangled.to_csv(os.path.join(path, 'prepared data', 'orders_clean.csv'))